lib/tapioca/dsl/compilers/rspec.rb

```diff
@@ -54,13 +71,400 @@ def create_includes(klass)
+            }
+            def make(mod:, path:, ast:, rsigs_procs:)
+              root = SemanticNode.make(ast:)
+              rsig_nodes = index_by(root.rsigs) { |n| n.location.begin.line }
+              methods_by_line = rsigs_procs.map do |sig_line_number, (sig_caller_location, decl_proc)|
+                target_sig_node = rsig_nodes[sig_line_number]
+                next unless target_sig_node
+
+                rsig_target = find_rsig_target(target_sig_node)
+                next unless rsig_target
+
+                target_line_number, target_node = rsig_target
+
+                rsig_target_generated_method_name =
+                  (target_node.method_name == :subject) ? :subject : target_node.sym_args[0]
+                next unless rsig_target_generated_method_name
+
+                method_semantics = MethodSemantics.new(
+                  method_name: rsig_target_generated_method_name,
+                  target_node:,
+                  target_sig_node:,
+                  target_line_number:,
+                  sig_line_number:,
+                  sig_caller_location:,
+                  decl_proc:
+                )
+                [target_line_number, method_semantics]
+              end.compact.to_h
+
+              new(mod:, path:, root:, methods_by_line:)
+            end
+          end
+        end
+
+        class ModuleSemantics < T::Struct
+          extend T::Sig
+          SemanticsCache = T.type_alias { T::Hash[String, FileSemantics] }
+          const :mod, Module
+          const :semantics, SemanticsCache, factory: -> { {} }
+
+          sig { params(unbound_method: UnboundMethod).returns(FileSemantics) }
+          def file_semantics(unbound_method)
+            path, = T.must(unbound_method.source_location)
+            return semantics.fetch(path) if semantics.key?(path)
+
+            buffer = Parser::Source::Buffer.new(path, source: File.read(path))
+            ast = T.let(::Parser.const_get(:CurrentRuby).new.parse(buffer), Parser::AST::Node)
+            rsigs_procs = file_rsigs(path)
+            semantics[path] = FileSemantics.make(mod:, path:, ast:, rsigs_procs:)
+          end
+
+          sig { returns(T::Hash[String, T::Hash[Integer, [Thread::Backtrace::Location, Proc]]]) }
+          def module_rsigs
+            return {} unless mod.instance_variable_defined?(:@__rsig_blocks)
+
+            T.cast(mod.instance_variable_get(:@__rsig_blocks),
+              T::Hash[String, T::Hash[Integer, [Thread::Backtrace::Location, Proc]]])
+          end
+
+          sig { params(file: String).returns(T::Hash[Integer, [Thread::Backtrace::Location, Proc]]) }
+          def file_rsigs(file) = module_rsigs[file] || {}
+
+          sig { params(unbound_method: UnboundMethod).returns(T.nilable(MethodSemantics)) }
+          def method_semantics(unbound_method) = file_semantics(unbound_method).method_semantics(unbound_method)
+        end
+
+        class CompilationWorkspace < T::Struct
```

Comment:
Is adding a new `CompilationWorkspace` nested class really necessary? It would seem to me that this code can just be part of `Tapioca::Dsl::Compilers::RSpec`.